protocol/consistencychecks.go

```diff
@@ -0,0 +1,328 @@
+}
+
+// verifySTR checks the consistency between 2 snapshots.
+// This uses the pinning signing key in cc
+// to verify the STR's signature and should not verify
+// the hash chain using the current STR stored in cc.
+func (cc *ConsistencyChecks) verifySTR(savedSTR, str *m.SignedTreeRoot) error {
+	// verify STR's signature
+	if !cc.signKey.Verify(str.Serialize(), str.Signature) {
+		return ErrorBadSignature
+	}
+
+	// verify hash chain
+	// TODO: clearly verify which is the client's actual expectation.
+	// See #81
+	if (str.Epoch == savedSTR.Epoch && bytes.Equal(savedSTR.Signature, str.Signature)) ||
+		(str.Epoch == savedSTR.Epoch+1 && str.VerifyHashChain(savedSTR.Signature)) {
+		return nil
+	}
+
+	// TODO: verify the directory's policies as well. See #115
+	return ErrorBadSTR
+}
+
+// verifyFulfilledPromise verifies issued TBs was inserted
+// in the directory as promised.
+func (cc *ConsistencyChecks) verifyFulfilledPromise(uname string,
+	str *m.SignedTreeRoot,
+	ap *m.AuthenticationPath) error {
+	if !cc.useTBs {
+		return nil
+	}
+
+	if cc.SavedSTR.Epoch+1 != str.Epoch {
+		return nil
+	}
+
+	tb := cc.TBs[uname]
+	if tb != nil {
+		if !tb.Verify(ap) {
+			return ErrorBrokenPromise
+		}
+		delete(cc.TBs, uname)
+	}
+	return nil
+}
+
+// verifyReturnedPromise validates the returned promise
+// based on the request type. Note that the directory
+// returns a promise iff the returned proof is
+// _a proof of absence_.
+// 	If the request is a registration, and
+// 	- the request is successful, then the directory should return a promise for the new binding.
+// 	- the request is failed because of ErrorNameExisted, then the directory should return a promise for that existed binding.
+//
+// 	If the request is a key lookup, and
+// 	- the request is successful, then the directory should return a promise for the lookup binding.
+func (cc *ConsistencyChecks) verifyReturnedPromise(requestType int,
+	responseCode ErrorCode, df *DirectoryProof,
+	uname string, key []byte) error {
+	if !cc.useTBs {
+		return nil
+	}
+
+	ap := df.AP
+	str := df.STR
+	tb := df.TB
+
+	// the client should receive a signed promise iff
+	// the directory returns a proof of absence
+	if bytes.Equal(ap.LookupIndex, ap.Leaf.Index) {
+		if tb != nil {
+			// the directory returned a proof of inclusion
+			// _and_ a promise
+			return ErrorMalformedDirectoryMessage
+		}
+		return nil
+	}
+
+	if tb == nil {
+		if (requestType == RegistrationType) ||
+			(requestType == KeyLookupType && responseCode == Success) {
+			return ErrorBadPromise
+		}
+		return nil
+	}
+
+	// verify TB's Signature
+	if !cc.signKey.Verify(tb.Serialize(str.Signature), tb.Signature) {
+		return ErrorBadSignature
+	}
+
+	// verify TB's VRF index
+	if !bytes.Equal(tb.Index, ap.LookupIndex) {
+		return ErrorBadIndex
+	}
+
+	// verify TB's value
+	// key could be nil if we may have no information about the existed binding
+	if key != nil && !bytes.Equal(tb.Value, key) {
+		return ErrorBadPromise
+	}
```

Comment:
> The last two if-cases need to be split up like this, to clearly state what went wrong (ErrorBadIndex/ErrorBadPromise)?

Yes, I'd like to split these checks as you said. Also, `tb.Verify` is being used to verify if the binding was inserted as promise.